src/main/java/com/google/devtools/build/lib/bazel/repository/CompressedTarFunction.java

```diff
@@ -54,20 +55,21 @@ public Path decompress(DecompressorDescriptor descriptor)
     Map<Path, PathFragment> symlinks = new HashMap<>();
 
     try (InputStream decompressorStream = getDecompressorStream(descriptor)) {
-      TarArchiveInputStream tarStream = new TarArchiveInputStream(decompressorStream);
+      // USTAR tar headers use an unspecified encoding whereas PAX tar headers always use UTF-8.
+      // Since this would result in ambiguity with Bazel forcing the default JVM encoding to be
+      // ISO-8859-1, we explicitly specify the encoding as UTF-8.
+      TarArchiveInputStream tarStream = new TarArchiveInputStream(decompressorStream, UTF_8.name());
```

Comment:
There doesn't seem to be an easy way to keep reading legacy tars in Latin-1 while still being able to distinguish them from UTF-8 names. 

In order to preserve backwards compatibility with Latin-1 tars without PAX headers, I *could* inject a custom `Charset` via `CharsetProvider` that somehow marks all strings it produces in a way that lets us distinguish them from those created by the default UTF-8 charset (e.g. via a `WeakIdentityHashMap`). Since that would definitely be a hack, I didn't want to do it without getting feedback on this point first.

Reply:
cc @tetromino (since he's been thnking about Unicode recently)

My professional opinion is that I don't know. What would his hack be useful for? I thought that what this change was doing was to parse the UTF-8 data into proper Unicode then convert that to Bazel's weird "UTF-8 bytes in ISO8859-1" encoding, but that seems doable without resorting to somehow tagging the `String`s produced by it by (essentially) doing "if valid UTF-8, decode it as such, otherwise go directly to the weird Bazel encoding".

Reply:
It's more messy than that: The Latin-1 trick only works if all strings are encoded in Latin-1 from source to sink, otherwise the "opaque bag of bytes" property is irrecoverably lost. 

But commons-compress breaks this invariant: While tar ustar headers respect the default string encoding (which can also be set via the constructor), the more modern tar pax headers always use UTF-8. 

Since Bazel can't determine which encoding was used to produce the file names returned by commons-compress, I found the only non-hacky option to be to force UTF-8 for ustar headers as well, decode UTF-8 and then reencode as Latin-1.

As a result, this change would break users of tars with non-UTF-8 ustar headers. Previously, these were handled correctly via the Latin-1 trick (but any pax header with non-ASCII chars were broken), but now they would be decoded as UTF-8 and thus turn into garbage.

Ideally we could tweak commons-compress to return byte arrays or at least tell us which type of header it parser, but I didn't find a way to do this. We could raise an issue with them, hack around this limitation by tracking the strings via a custom charset or accept that life isn't fair and UTF-8 deserves to be the winner. ;⁠-⁠)

Reply:
Well, actually, it's even more messy than that, if I understand correctly: the nice thing about the Latin-1 hack is that it works as long as the encoding on the input is the same as the encoding on the output. However, apparently, pax headers insist that they are UTF-8, which means that they constrain the output of Bazel  to UTF-8 also. 

Now I see where you are coming from: commons-compress doesn't tell whether there are pax headers and thus it doesn't tell you whether UTF-8 decoding was required. So what you could do is to keep track of `String`s that were emitted by the `Charset`, thus you know which `String`s went through the UTF-8 decoding and thus undo it. Is this correct?

I don't claim any knowledge of either commons-compress or the intricacies of tar files, but `TarArchiveEntry` seems to have some methods that have `Pax` in their name. Maybe those could be used to disambiguate instead of the hack you proposed?

Reply:
As far as I can tell you can get your hands on *extra* pax headers, but not on the standard ones including the file name. It's just silently updated to the pax header value when encountered.

Reply:
Note that https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html#tag_20_92_13_03 states that pax headers can be UTF-8 or "binary" (OS-specified encoding).

Is the "binary" option never used in practice?

Reply:
@tetromino Not in commons-compress, as far as I can tell: https://github.com/apache/commons-compress/blob/3605bc0e27c1f1d27868e7ad5665513835ca7bec/src/main/java/org/apache/commons/compress/archivers/tar/TarUtils.java#L799

Reply:
@tetromino WDYT about the above hack with a custom `Charset` @fmeum proposed?

I am kinda drawn to it, because it wouldn't affect anything outside the interpretation of that particular `.tar` file and would allow us to work around the above limitation, but it's messy enough that I don't want to bless it without a second opinion.

Reply:
@tetromino Friendly ping, let me know whether you want me to implement the hack.

Reply:
@tetromino Friendly ping

Reply:
@lberki Since I wanted to know whether the hack would work, I went ahead and implemented it. It does seem to work and I didn't even have to use a weak cache.

Reply:
@lberki @tetromino Friendly ping, I would really like to get this into Bazel 7. We currently have to maintain a number of hacks in rules_go just to be able to extract a Go SDK.

Reply:
My gut reaction to the encoding hack was pretty negative. Then I reread the code and description a couple times, and decided that the hack is only ugly because the apache api is insufficiently flexible for bad clients like us that want to do weird things with encodings. I'm ok with the hack provided it is sufficiently explained in comments, and doesn't leak outside this file.

I would like us to be pretty explicit in comments about what the format of the input headers and returned string are. I.e., Javadoc for the class `CompressedTarFunction` or its `decompress()` method, saying that USTAR headers are interpreted as latin-1, PAX headers are interpetted as UTF-8, and both types are turned into Java Strings holding the UTF-8 code units that represent the string content, in order to match our string behavior elsewhere.

One more question. We eventually want to fix Bazel's decoding of Starlark source files to be UTF-8 instead of latin-1. Does anything in this PR make that harder? It seems to me this PR is just preserving an existing behavior (USTAR) while fixing an erroneous one (PAX).

Reply:
I added more extensive comments, including Javadocs for the class.

> One more question. We eventually want to fix Bazel's decoding of Starlark source files to be UTF-8 instead of latin-1. Does anything in this PR make that harder? It seems to me this PR is just preserving an existing behavior (USTAR) while fixing an erroneous one (PAX).

I don't think that it would make this harder. It does introduce one or two new places in which `ISO_8859_1` would need to be replaced with `UTF-8`, but those are only related to prefix stripping, not the actual entry decoding logic that interfaces with commons-compress and relies on the custom decoder hack.